Add fallback for the default font folder path

matiasbenedetto opened the pull request:
## What?

Add a fallback for the default font directory.
Non persisted alternative of: https://github.com/WordPress/wordpress-develop/pull/6252


## Why?
> For installations that don’t support modification of the wp-content directory, the Font Library will use wp-content/uploads/fonts as a fallback location, ensuring we stay true to our project philosophy of designing for the majority while still making the feature available to anyone out of the box without extra steps from the user. 

Decision from https://make.wordpress.org/core/2024/03/07/unblocking-wp6-5-font-library-and-synced-pattern-overrides/
Discussion: https://github.com/WordPress/gutenberg/issues/59699

## How?
Update the `wp_get_font_dir()` function.

### Default
**If `/wp-content/fonts` is writable**,  `wp_get_font_dir()` should return:
```php
array(
    'path' => '/var/www/src/wp-content/fonts',
    'url' => 'http://localhost:8889/wp-content/fonts',
    'subdir' => '',
    'basedir' => '/var/www/src/wp-content/fonts',
    'baseurl' => 'http://localhost:8889/wp-content/fonts',
    'error' => false
)
```

### Fallback
**If `/wp-content/fonts` is not writable**,  `wp_get_font_dir()` should return:
```php
array(
    'path' => '/var/www/src/wp-content/uploads/fonts',
    'url' => 'http://localhost:8889/wp-content/uploads/fonts',
    'subdir' => '',
    'basedir' => '/var/www/src/wp-content/uploads/fonts',
    'baseurl' => 'http://localhost:8889/wp-content/uploads/fonts',
    'error' => false
)
```
**If the `wp-content/uploads/fonts` can not be created or written**, an error message is added in the 'error' key of the returned array following the `wp_font_dir` pattern: https://github.com/matiasbenedetto/wordpress-develop/blob/trunk/src/wp-includes/functions.php#L2393-L2406

```php
array(
    'path' => '/var/www/src/wp-content/uploads/fonts',
    'url' => 'http://localhost:8889/wp-content/uploads/fonts',
    'subdir' => '',
    'basedir' => '/var/www/src/wp-content/uploads/fonts',
    'baseurl' => 'http://localhost:8889/wp-content/uploads/fonts',
    'error' => 'Unable to create directory wp-content/uploads/fonts. Is its parent directory writable by the server?'
)
```

### Using `font_dir` filter

Example filtering:
```php
// Define a callback function to pass to the filter.
function set_custom_dir( $defaults ) {
	$defaults['path']    = path_join( WP_CONTENT_DIR, 'custom_dir' );
	$defaults['url']     = 'http://example.com/custom-path/fonts/my-custom-dir';
	$defaults['basedir'] = path_join( WP_CONTENT_DIR, 'custom_dir' );
	$defaults['baseurl'] = 'http://example.com/custom-path/fonts/my-custom-dir';
	return $defaults;
}

// Add the filter.
add_filter( 'font_dir', 'set_custom_dir' );
```

**If `font_dir` is filtered with a writable directory**, `wp_get_font_dir()` should return:

```php
array(
    'path' => '/var/www/src/wp-content/uploads/custom_dir',
    'url' => 'http://example.com/custom-path/fonts/my-custom-dir',
    'subdir' => '',
    'basedir' => '/var/www/src/wp-content/uploads/custom_dir',
    'baseurl' => 'http://example.com/custom-path/fonts/my-custom-dir',
    'error' => ''
)
```

**If `font_dir` is filtered with a NON writable directory**, `wp_get_font_dir()` should return:

```php
array(
    'path' => '/var/www/src/wp-content/uploads/custom_dir',
    'url' => 'http://example.com/custom-path/fonts/my-custom-dir',
    'subdir' => '',
    'basedir' => '/var/www/src/wp-content/uploads/custom_dir',
    'baseurl' => 'http://example.com/custom-path/fonts/my-custom-dir',
    'error' => 'Unable to create directory wp-content/custom_dir. Is its parent directory writable by the server?'
)
```
---

Trac ticket: https://core.trac.wordpress.org/ticket/60751#ticket

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

matiasbenedetto committed:
Add fallback for the default font folder path

matiasbenedetto committed:
improve dir is_writable check

matiasbenedetto committed:
format php

matiasbenedetto committed:
remove unwanted early return

matiasbenedetto committed:
re-order actual and expected

matiasbenedetto committed:
remove not needed assignement

matiasbenedetto committed:
do not suppres errors

matiasbenedetto committed:
format php

matiasbenedetto committed:
use the same format as in the wp-content/fonts default path

matiasbenedetto committed:
removing database persistence

matiasbenedetto committed:
remove wp_default_font_dir function as not needed

peterwilsoncc commented on src/wp-includes/fonts.php:
```diff
@@ -122,6 +122,17 @@ function wp_get_font_dir() {
 		'error'   => false,
 	);
 
+	wp_mkdir_p( $defaults['path'] );
```
```suggestion
	$made_dir = wp_mkdir_p( $defaults['path'] );
```

Good catch, it will need to exist in order to check it's writable.

peterwilsoncc commented on src/wp-includes/fonts.php:
```diff
@@ -122,6 +122,17 @@ function wp_get_font_dir() {
 
+	wp_mkdir_p( $defaults['path'] );
+
+	if ( ! wp_is_writable( $defaults['path'] ) ) {
```
```suggestion
	if ( ! $made_dir || ! wp_is_writable( $defaults['path'] ) ) {
```

`wp_mkdir_p()` returns true if the directory exists or could be created. If it returns false, then it's already known that the path isn't writable.

peterwilsoncc commented on src/wp-includes/fonts.php:
```diff
@@ -122,6 +122,17 @@ function wp_get_font_dir() {
+	wp_mkdir_p( $defaults['path'] );
+
+	if ( ! wp_is_writable( $defaults['path'] ) ) {
+		$defaults            = wp_upload_dir();
```
```suggestion
		$defaults            = wp_get_upload_dir();
```

This prevents the time based folder from being created, eg `uploads/20204/03`, if it doesn't already exist.

I think this will require https://github.com/WordPress/wordpress-develop/pull/6211 to prevent a loop. Are you able to add a test (there's one in the linked PR).

matiasbenedetto replied:
added wp_get_upload_dir here https://github.com/WordPress/wordpress-develop/pull/6259/commits/895984bdf4c441d5369bb490e172c088077b3208

peterwilsoncc commented on src/wp-includes/fonts.php:
```diff
@@ -122,6 +122,17 @@ function wp_get_font_dir() {
+
+	if ( ! wp_is_writable( $defaults['path'] ) ) {
+		$defaults            = wp_upload_dir();
+		$defaults['path']    = path_join( $defaults['basedir'], 'fonts' );
+		$defaults['url']     = $defaults['baseurl'] . '/fonts';
+		$defaults['subdir']  = '';
+		$defaults['basedir'] = path_join( $defaults['basedir'], 'fonts' );
+		$defaults['baseurl'] = $defaults['baseurl'] . '/fonts';
```
Nit: defaults can simply be redefined, ie `$defaults = array( /* etc */ );`

matiasbenedetto replied:
done here: https://github.com/WordPress/wordpress-develop/pull/6259/commits/9b05b6a5dd1a8cb94acbd045995ca6e0412722de

peterwilsoncc reviewed (commented):
I've added a few notes inline.

There will need to be some tests to ensure against:

* infinite loops (as noted)
* fallback applies when expected
* fallback does not apply when it isn't expected

If neither directory can be created then `error` should be defined and return a similar result to `wp_upload_dir()`.

matiasbenedetto committed:
save wp_mkdir_p result in a variable

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

matiasbenedetto committed:
check if the dir couldn't be created

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

matiasbenedetto committed:
use array syntax

matiasbenedetto committed:
Update wp_get_font_dir function to use wp_get_upload_dir instead of wp_upload_dir

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

matiasbenedetto commented:
> There will need to be some tests to ensure against:
>     * infinite loops (as noted)
>     * fallback applies when expected
>     * fallback does not apply when it isn't expected
> If neither directory can be created then `error` should be defined and return a similar result to `wp_upload_dir()`.

I thought about these 2 alternatives for testing this functionality but they didn't work on my end:
1. use `chmod()` in the test body to change the permissions of the path (it seems to be ignored, and wp_is_writable always returns true in the unit test env).
2. try to mock the `wp_is_writable( $path )` function in the context of the unit test. (I couldn't mock this global function. I found examples in the core's codebase about mocking class methods but not functions in the global scope).

Any help bout how to test this functionality is welcome.


cc. @peterwilsoncc @swissspidy @youknowriad

matiasbenedetto committed:
If the `wp-content/uploads/fonts` can not be created or written, an error message is added in the 'error' key of the returned array following the `wp_font_dir` pattern: https://github.com/matiasbenedetto/wordpress-develop/blob/trunk/src/wp-includes/functions.php#L2393-L2406

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

matiasbenedetto commented:
> If neither directory can be created then error should be defined and return a similar result to wp_upload_dir().

Added here: https://github.com/WordPress/wordpress-develop/pull/6259/commits/d2f6e5c03ef6ed041a7dfb6f71ecfda4dce3504b

matiasbenedetto committed:
be consistent in the use of $defaults['path'] to determine the path

swissspidy commented:
@matiasbenedetto Which test is this in context of? This PR does not have any tests, but I do see tests at #6252. The `chmod()` there actually works just fine for me.

Some of my recent l10n tests also use `chmod` without issues, so I don't see why that would be an issue here.

If you can point me at the code this is in reference to, I am happy to take a look.

matiasbenedetto committed:
add a test using chmod (is failing)

matiasbenedetto committed:
format php

matiasbenedetto commented:
> @matiasbenedetto Which test is this in context of? This PR does not have any tests, but I do see tests at #6252. The `chmod()` there actually works just fine for me.


@swissspidy thanks for taking a look. The test using `chmod` is failing in #6252 CI: https://github.com/WordPress/wordpress-develop/actions/runs/8251059911/job/22567291729?pr=6252

![image](https://github.com/WordPress/wordpress-develop/assets/1310626/c95ead26-44de-4a3b-8ec0-4f95f13787b7)



> Some of my recent l10n tests also use `chmod` without issues, so I don't see why that would be an issue here.
> 
> If you can point me at the code this is in reference to, I am happy to take a look.

I added the same kind of test using `chmod` in this PR to test here too but is failing on my local env and in the CI: https://github.com/WordPress/wordpress-develop/pull/6259/commits/9c51258044730882a6e1e50e747882a16869f02e

matiasbenedetto committed:
undo unwanted changes

matiasbenedetto committed:
fix test

matiasbenedetto commented on tests/phpunit/tests/fonts/font-library/wpFontsDir.php:
```diff
@@ -69,4 +69,25 @@ function set_new_values( $defaults ) {
+
+	public function test_fonts_dir_with_uploads_default() {
+		$font_dir = wp_get_font_dir();
+		chmod( $font_dir['path'], 0000 );
```
This line is not working as expected. The `/wp-content/fonts` dir continues being evaluated as writable by the `wp_is_writable()` function.

hellofromtonya replied:
Changing directory permissions isn't going to work. 

* `0000` will not be resettable during tear down, as it'll throw a permissions denied error.
* Setting `wp-content` to read-only => `is_rewritable()` returns `true`.
* Adding an existing `wp-content/fonts` and setting it to read-only (before testing) => `is_rewritable()` returns `true`.

hellofromtonya replied:
@costdev had an idea to create a new `wp-content/fonts` file (not directory) as a fake to trigger the fallback logic. It works. It's weird. But it works for what needs to get tested.

hellofromtonya replied:
Updated to use this approach https://github.com/WordPress/wordpress-develop/pull/6259/commits/5f7b3d1bf2adf9b62165d7208e0108f80120dff4 ✅

swissspidy committed:
Make test more robust

swissspidy commented:
@matiasbenedetto The test needs to be more robust and properly cleanup during teardown. Just pushed a first step towards this.

swissspidy committed:
Delete files in directory too

peterwilsoncc commented on src/wp-includes/fonts.php:
```diff
@@ -122,6 +122,36 @@ function wp_get_font_dir() {
 		'error'   => false,
 	);
 
+	$made_dir = wp_mkdir_p( $defaults['path'] );
```
I've come to realize that attempting to create the directory here is problematic.

As the code runs before the `font_dir` filter, it's unknown  at this point in the execution whether the directory path will be overridden by a plugin. If it is overridden then this code may create a junk directory in the sites filesystem that will never be used. 

What I think the code will need to do is:

* check if the folder exists and is writable
* if the folder doesn't exist, check the parent exists and is writable
* rely on `wp_upload_dir()` to create the directory after the `upload_dir` filter runs (which will include firing the `font_dir` filter) 
* check the error status produced by `wp_upload_dir()`

azaozz replied:
> create the directory after the upload_dir filter runs (which will include firing the font_dir filter)

+1. 

Seems the first two checks:
- check if the folder exists and is writable
- if not, check the parent exists and is writable

are [also done in `wp_mkdir_p()`](https://github.com/WordPress/wordpress-develop/blob/6.4/src/wp-includes/functions.php#L2034-L2099). Somewhat unsure if they need to be done (repeated) before running wp_mkdir_p()?

hellofromtonya replied:
Yes, good points @peterwilsoncc. I agree 👍

matiasbenedetto replied:
Good call; in [this commit](https://github.com/WordPress/wordpress-develop/pull/6259/commits/90a3982636f1e47209e19a8a170ce027fb921af0) I implemented the change to create the folder only after the `font_dir` filter runs.

hellofromtonya committed:
Fake a failure for creating wp-content/fonts directory.

To trigger the fallback logic, need to create a fake where `wp_mkdir_p()` bails out early and returns `false`.

Looking at wp_mkdir_p(), notice this code:

```
if ( file_exists( $target ) ) {
	return @is_dir( $target );
}
```

When the `$target` exists but is a file (not a directory), then `false` gets returned. Aha! Props to @costdev.

This commit:

* Adds a fake file, reworks the test, and then removes the file in the tear_down().
* Refactors the removing of the fonts directory by putting it into a separate method.

These new methods can be put into a trait or base abstraction that all of the Font Library test classes use to ensure the filesystem is restored between tests (that's out-of-scope for this PR).

hellofromtonya commented:
Whomever commits this, please also prop @costdev.

hellofromtonya commented on tests/phpunit/tests/fonts/font-library/wpFontsDir.php:
```diff
@@ -69,4 +81,57 @@ function set_new_values( $defaults ) {
+		}
+	}
+
+	private function fake_no_new_directories_in_wp_content() {
```
This is a weird fake, but it works. The way it works is this:

A file named `fonts` that exists in the `wp-content` will trigger `wp_mkdir_p()` to fail into the `file_exists()` bail out and return false when `is_dir()` runs.

That will then cause the fallback branch to execute.

hellofromtonya replied:
I created it in a method for 2 reasons:

1. In the test, makes it more readable and less code.
2. Reusability. This logic might be useful elsewhere for other Font Library tests. If yes, it can be moved to a trait or base TestCase.

hellofromtonya replied:
Wondering if the method name should be focused on what it does -> makes `wp_mkdir_p()` return `false` ??

swissspidy replied:
I think the name is fine

matiasbenedetto replied:
I renamed this function `create_fake_file_to_avoid_dir_creation` because I think it's a little clearer, making it path-agnostic. It was needed because it is used to create a path received as a parameter in the additional test cases added here: https://github.com/WordPress/wordpress-develop/pull/6259/commits/90a3982636f1e47209e19a8a170ce027fb921af0

hellofromtonya commented on tests/phpunit/tests/fonts/font-library/wpFontsDir.php:
```diff
@@ -69,4 +81,57 @@ function set_new_values( $defaults ) {
+		$this->assertSame( $expected, $font_dir, 'The font directory should be a subdir in the uploads directory.' );
+	}
+
+	private function remove_fonts_directory() {
```
I moved the cleanup code to a separate method to make it reusable.

All of the Font Library tests should be removing the fonts directory between tests. Moving it here is the step. Then a trait or base TestCase can be add that is shared amongst all of the test classes (though out-of-scope for this PR).

ironprogrammer commented on tests/phpunit/tests/fonts/font-library/wpFontsDir.php:
```diff
@@ -69,4 +81,57 @@ function set_new_values( $defaults ) {
 		$this->assertSame( static::$dir_defaults, $font_dir, 'The wp_get_font_dir() method should return the default values.' );
 	}
+
+	public function test_should_created_fonts_dir_in_uploads_when_fails_in_wp_content() {
```
```suggestion
	public function test_should_create_fonts_dir_in_uploads_when_fails_in_wp_content() {
```
Fixes small typo.

Would `test_should_create_fonts_dir_in_uploads_when_wp_content_not_writable` be clearer here?

ironprogrammer commented on tests/phpunit/tests/fonts/font-library/wpFontsDir.php:
```diff
@@ -69,4 +81,57 @@ function set_new_values( $defaults ) {
+		}
+	}
+
+	private function fake_no_new_directories_in_wp_content() {
```
```suggestion
	/**
	 * A placeholder "fake" file at `wp-content/fonts` triggers `wp_mkdir_p()` to fail into the `file_exists()` bail out, causing `is_dir()` to  return `false`.
	 * This effectively makes `wp-content/fonts` unwritable as the default `fonts` directory location.
	 */
	private function fake_no_new_directories_in_wp_content() {
```
Added a summary of discussion around this approach for future reference (thanks, @hellofromtonya 🙌🏻). It's awkward and could use some massaging if anyone has suggestions! 🙏🏻

costdev replied:
How about just saying something like this?

```
Creates a file with the same name as the directory. This means the directory cannot be created.
```

That way, it's implementation-independent for possible reuse in the future. Alternatively, a similar description that doesn't go quite so much into the implementation details would work here too. Just a thought, not a blocker.

ironprogrammer commented on tests/phpunit/tests/fonts/font-library/wpFontsDir.php:
```diff
@@ -69,4 +81,57 @@ function set_new_values( $defaults ) {
+		$font_dir = wp_get_font_dir();
+
+		$this->assertDirectoryDoesNotExist( path_join( WP_CONTENT_DIR, 'fonts' ), 'The `wp-content/fonts` directory should not exist' );
+		$this->assertDirectoryExists( $font_dir['path'], 'The `uploads/fonts` directory should exist' );
```
```suggestion
		$this->assertDirectoryDoesNotExist( path_join( WP_CONTENT_DIR, 'fonts' ), 'The `wp-content/fonts` directory should not exist.' );
		$this->assertDirectoryExists( $font_dir['path'], 'The `uploads/fonts` directory should exist.' );
```
Adds periods to `$message` for consistency.

ironprogrammer commented on tests/phpunit/tests/fonts/font-library/wpFontsDir.php:
```diff
@@ -69,4 +81,57 @@ function set_new_values( $defaults ) {
+
+		$this->assertDirectoryDoesNotExist( path_join( WP_CONTENT_DIR, 'fonts' ), 'The `wp-content/fonts` directory should not exist' );
+		$this->assertDirectoryExists( $font_dir['path'], 'The `uploads/fonts` directory should exist' );
+		$this->assertSame( $expected, $font_dir, 'The font directory should be a subdir in the uploads directory.' );
```
```suggestion
		$this->assertSame( $expected, $font_dir, 'The `fonts` directory should be a subdir in the `uploads` directory.' );
```
Nits for consistency.

ironprogrammer reviewed (commented):
Thanks for this PR, and for everyone's input in fleshing it out!

In addition to the suggestions in this review, I also recommend adding the following annotations to the file header:

```
@since 6.5.0
@ticket 60751
```

And while not added in this PR, the function `test_fonts_dir_with_filter()` might be better named `test_font_dir_filter()`, as the filter name is `font_dir`.

matiasbenedetto committed:
Improve texts

Co-authored-by: Brian Alexander <824344+ironprogrammer@users.noreply.github.com>

matiasbenedetto committed:
improve test message

Co-authored-by: Brian Alexander <824344+ironprogrammer@users.noreply.github.com>

swissspidy commented on tests/phpunit/tests/fonts/font-library/wpFontsDir.php:
```diff
@@ -69,4 +81,61 @@ function set_new_values( $defaults ) {
+
+	private function remove_no_new_directories_in_wp_content_fake() {
+		if ( file_exists( self::$fake_fonts_file ) ) {
+			@unlink( self::$fake_fonts_file );
```
I don't think we need or should add error suppression.

matiasbenedetto replied:
removed here: https://github.com/WordPress/wordpress-develop/pull/6259/commits/f7de278c467ffa43205e32f0605aee6b94d988fd

matiasbenedetto committed:
If the font_dir was filtered it uses the filtered values and it doesn't fall back to the /wp-content/uploads/fonts folder.

matiasbenedetto committed:
remove error supression

matiasbenedetto committed:
format php

matiasbenedetto committed:
rename test case

matiasbenedetto commented:
> the function test_fonts_dir_with_filter() might be better named test_font_dir_filter(), as the filter name is font_dir.

Done here: https://github.com/WordPress/wordpress-develop/pull/6259/commits/f0c3f3dfb3bfe0aa02352e55932c754458130ca7

matiasbenedetto commented:
Thanks everyone for the contributions.

In the last commits, I tried to implement the remaining feedback.
I updated the PR description with all the cases being tested and their expected returned values.
I think it is ready for another review.

peterwilsoncc commented on src/wp-includes/fonts.php:
```diff
@@ -131,7 +131,44 @@ function wp_get_font_dir() {
 	 * @param array $defaults The original fonts directory data.
 	 */
-	return apply_filters( 'font_dir', $defaults );
+	$font_dir = apply_filters( 'font_dir', $default_dir );
```
This needs to be applied with the final default directory known.

While the use-case we've seen is to change the default directory, there are other use cases that need to be considered. Once such example is wishing to store different font families in sub-directories, eg fonts/open-sans, fonts/source-code-pro.

It also presents the problem that WordPress may override the changes introduced by a plugin. If the directory defined by the plugin is not writable then it's the plugin's problem to handle.

hellofromtonya replied:
+1 I agree. 

The processing order then:

1. First, determine _where_ the `fonts` directory should be created.
2. Then pass that "final" location to the filter.
3. Then use that filtered location to make the directory.

hellofromtonya replied:
As @peterwilsoncc noted, the first step of determining the "where" needs to include more conditionals, [such as checking if the parent is writable](https://github.com/WordPress/wordpress-develop/pull/6259/files#r1525500805).

peterwilsoncc commented on src/wp-includes/fonts.php:
```diff
@@ -131,7 +131,44 @@ function wp_get_font_dir() {
-	return apply_filters( 'font_dir', $defaults );
+	$font_dir = apply_filters( 'font_dir', $default_dir );
+
+	wp_mkdir_p( $font_dir['path'] );
```
This function can not attempt to create the directory. It's unaware if `wp_upload_dir()` has been called to get the directory without creating it.

If the directory doesn't exist, we'll need to check if the parent directory is writable.

azaozz replied:
As far as I see `wp_mkdir_p()` expects full paths and makes directories recursively (calls PHP's `mkdir()` with `true` as third param). That should make any missing parent and sub-parent directories? I.e. trying to make `/wp-content/dir1/dir2/fonts` would work, and `dir1` and `dir2` will be created if they don't exist.

> If the directory doesn't exist, we'll need to check if the parent directory is writable.

May be missing something but seems `wp_mkdir_p()` would return `false` if the parent (or grandparent, great-grandparent, etc.) directory is not writable. Seems that's how it is used elsewhere in core? Then checking beforehand whether the parent directory is writable sounds like a tiny bit of optimization for the cases where that directory is not writable, but a tiny bit of overhead for cases where it is writable (more common)? 

I.e. something like this code (example copied from [wp_mkdir_p()](https://github.com/WordPress/wordpress-develop/blob/6.4/src/wp-includes/functions.php#L2034-L2099)) will have to run before every `wp_mkdir_p()` call:
```
// Do not allow path traversals.
if ( str_contains( $target, '../' ) || str_contains( $target, '..' . DIRECTORY_SEPARATOR ) ) {
	return false; // Bail out somehow...
}

// We need to find the permissions of the parent folder that exists and inherit that.
$target_parent = dirname( $target );
while ( '.' !== $target_parent && ! is_dir( $target_parent ) && dirname( $target_parent ) !== $target_parent ) {
	$target_parent = dirname( $target_parent );
}

// Get the permission bits.
$stat = @stat( $target_parent );
// Or maybe use is_writable( $target_parent ); (seems this might be a bit problematic on some filesystems if $target_parent was just created)
```

And then almost the same code will run inside `wp_mkdir_p()` if the permissions indicate that the directory is writable.

Frankly I'm not sure that the extra code to check the parent directory's permissions is warranted? Seems `wp_mkdir_p()` can handle the case where parent is not writable?

peterwilsoncc replied:
> Frankly I'm not sure that the extra code to check the parent directory's permissions is warranted? Seems `wp_mkdir_p()` can handle the case where parent is not writable?

`wp_upload_dir()` (which this code filters) runs in two context: one in which it creates the directory, one in which it doesn't.

As this function doesn't know the context re file creation, it's not able to create a directory as it risks adding junk directories to the server. Junk meaning that they may never be used.

The font directory functions are also used as a getter during post deletion so we can't create directories to delete files that do not exist.

azaozz replied:
Sorry but I think I'm really missing something here.

> wp_upload_dir() (which this code filters) runs in two context: one in which it creates the directory, one in which it doesn't.

Right.

> As this function doesn't know the context re file creation

I.e. we don't know whether `wp-content` or `wp-content/uploads` (or their replacements if somebody moved/renamed them) exist, correct? But how could that be after `wp_upload_dir()` has already returned the path and URL to the uploads dir? Do you mean the data from `wp_upload_dir()` may be wrong?

> it's not able to create a directory as it risks adding junk directories to the server

Meaning `wp_mkdir_p()` (actually PHP's `mkdir()`) will not be able to create a directory but may add "junk" directories? Sorry but I'm lost, how can it add directories without being able to create them...

Perhaps some practical example would help? Also, what would a fix for this look like? Maybe that would help to understand and recreate the problem?

peterwilsoncc replied:
It has nothing to do with whether `wp_mkdir_p()` can create the directory.

What I want to avoid is this function attempting to create the directories when it should not. This can happen in two circumstances:

* when hooked on to `upload_dir`: [`wp_upload_dir()`](https://developer.wordpress.org/reference/functions/wp_upload_dir/) is called with the `$create_dir` parameter set to false, either directly or via [`wp_get_upload_dir()`](https://developer.wordpress.org/reference/functions/wp_get_upload_dir/).
* when called as a getter in [`_wp_before_delete_font_face()`](https://github.com/WordPress/wordpress-develop/blob/9c415ed5fb9b4069208b9e43ff774acbfb3ef3a6/src/wp-includes/fonts.php#L163-L183): this is used to delete font files if the `_wp_font_face_file` post meta is defined. As font face post types may not contain a font file (see the discussion in https://github.com/WordPress/gutenberg/issues/59387) then WordPress can't start creating directories in order to delete files that don't exist.

azaozz replied:
Ah I see what you mean. So basically this function has to be split in two: a getter and then a function that would only be used to create the dirs if needed.

peterwilsoncc replied:
@azaozz Yes, that's correct. 

It's what I did in the PR to fix the infinite loop issue rather than work aroudn it https://github.com/WordPress/wordpress-develop/pull/6211 (reviews welcome, btw), partly to prepare for this.

peterwilsoncc commented on src/wp-includes/fonts.php:
```diff
@@ -131,7 +131,44 @@ function wp_get_font_dir() {
+			'error'   => false,
+		);
+
+		wp_mkdir_p( $font_dir['path'] );
```
As above

azaozz commented on src/wp-includes/fonts.php:
```diff
@@ -131,7 +131,44 @@ function wp_get_font_dir() {
+			'baseurl' => $upload_dir['baseurl'] . '/fonts',
+			'error'   => false,
+		);
+
```
As @peterwilsoncc mentions above thinking it would probably be good to run the `font_dir` filter here again as the values were reset (hardcoded).

peterwilsoncc replied:
I'd really prefer to just run it once after WordPress figures out what it will present as the default.

azaozz replied:
If I'm understanding correctly you prefer this function to do something like this?
```
if ( is_dir( WP_CONTENT_DIR ) && is_writable( WP_CONTENT_DIR ) ) {
    $default_font_dir = WP_CONTENT_DIR . '/fonts';
elseif ( is_dir( WP_CONTENT_DIR . '/uploads' ) && is_writable( WP_CONTENT_DIR . '/uploads' ) ) {
    // This will only work if /uploads already exists. The assumption is that /uploads cannot be created here as WP_CONTENT_DIR either doesn't exist or is not writable.
    $default_font_dir = WP_CONTENT_DIR . '/uploads/fonts';
} else {
    // Throw error, etc.
}

$font_dir = apply_filters( 'font_dir', $default_font_dir );

// Try to crate the /fonts dir
```

peterwilsoncc commented on tests/phpunit/tests/fonts/font-library/wpFontsDir.php:
```diff
@@ -69,4 +76,113 @@ function set_new_values( $defaults ) {
+		$this->assertSame( $expected, $font_dir, 'As /wp-content/uplods/fonts is not writable the error key should be populated with an error message.' );
+	}
+
+	private function remove_font_paths() {
+		$paths = array(
+			path_join( WP_CONTENT_DIR, 'fonts' ),
+			path_join( WP_CONTENT_DIR, 'custom_dir' ),
+			path_join( WP_CONTENT_DIR, 'uploads/fonts' ),
+		);
+
+		foreach ( $paths as $path ) {
+			if ( ! is_dir( $path ) ) {
+				if ( file_exists( $path ) ) {
+					unlink( $path );
+				}
+			} else {
+				$this->rmdir( $path );
+				rmdir( $path );
+			}
+		}
+	}
+
+	/**
+	 * A placeholder "fake" file at $path triggers `wp_mkdir_p()` to fail into the `file_exists()` bail out, causing `is_dir()` to  return `false`.
+	 * This effectively makes $path unwritable.
+	 */
+	private function create_fake_file_to_avoid_dir_creation( $path ) {
+		file_put_contents(
+			$path,
+			'fake file'
+		);
+	}
```
These two methods will cause problems for most local development environments, including the docker container provided in this repo.

Typically the test suite and the development environment share content directories so developers running these environments may have existing font folders as a result of their manual testing.

During my tests the fonts from manual testing were deleted during these test runs. 

To reproduce using the repos docker environment:

1. Run commands to bootstrap the environment
   ```
   npm install
   npm run build:dev
   npm run env:start
   npm run env:install
   ```
2. Visit http://localhost:8889/wp-admin/site-editor.php?canvas=edit
3. Opens Styles > Typography > Manager > Install Fonts & Install some fonts via google
4. Apply the font to the main header in 2024
5. Save and exit the site editor
6. Observe the folder `src/wp-content/fonts` has been created
7. Run this test suite `npm run test:php -- --filter Tests_Fonts_WpFontDir`
8. Observe the folder `src/wp-content/fonts` has been deleted
9. Visit the front end of the site 
10. Observe the font download fails with a 404 error

matiasbenedetto replied:
> Typically the test suite and the development environment share content directories so developers running these environments may have existing font folders as a result of their manual testing.

I think this is unfortunate and unexpected, right? I get your point but I think this is more a problem of the core dev env than a problem of the tests.

The unit tests are supposed to run in a reproducible environment and not depend on or produce 'side effects.' If other parts of the workflow unrelated to unit tests are affected by the changes made by running the unit tests, the workflow is deficient, and it should be improved instead of degrading the unit tests.

hellofromtonya replied:
You're right @peterwilsoncc. I think what is needed is to:

* Check if the `fonts` folder already exists (do that within the `set_up_before_class()` fixture).
* If no, do nothing.
* If yes,
    * Store it in a static property to remember. 
    * If that property is set, then disable deleting the existing folder (i.e. in the tear down) and skip the associated tests that will fail due to it already existing.
    
In Core's CI, the tests will run as the folder will not exists. But in local development environments, this technique will guard against the scenario you described.

azaozz replied:
> Check if the fonts folder already exists

+1. If the /fonts dir already exists (at any tested location) it can be used for testing but shouldn't be deleted. As far as I see keeping a static var whether the fonts dir pre-existed would be a good way to do this.

costdev replied:
The approach seems sound, though I'd do the same for `wp-content/uploads/fonts`. This may also already exist for manual testing purposes, and the test suite shouldn't delete it.

Ideally, we would be using a tests-only fonts directory to avoid affecting the actual site's filesystem in the tests and also preventing possible filename clashes, etc. Alternatively, if we continue to use the actual site's filesystem, maybe store the existence _and_ pre-test contents of the directories. This means we could accurately compare the contents before/after the function under test has run, and therefore detect unexpected creation/deletion of files in those directories, plus delete only the files created by the tests.

peterwilsoncc commented on tests/phpunit/tests/fonts/font-library/wpFontsDir.php:
```diff
@@ -16,30 +16,37 @@ class Tests_Fonts_WpFontDir extends WP_UnitTestCase {
 	}
 
-	public function test_fonts_dir_with_filter() {
+	public function test_font_dir_filter() {
```
I've realised this will trigger test failures on some of the hosting providers running the [test suite](https://make.wordpress.org/hosting/test-results/) if the content folder isn't writable on their file system.

creativecoder commented on src/wp-includes/fonts.php:
```diff
@@ -131,7 +131,44 @@ function wp_get_font_dir() {
+
+		// If the default directory was filtered it uses the filtered values and doesnt fall back to the wp-content/uploads/fonts.
+		if ( $font_dir !== $default_dir ) {
+			$font_dir['error'] = sprintf(
+				/* translators: %s: Directory path. */
+				__( 'Unable to create directory %s. Is its parent directory writable by the server?' ),
+				esc_html( str_replace( ABSPATH, '', $font_dir['path'] ) )
+			);
```
Note that there are a few places where `wp_get_font_dir` is used that we'll need to check to see if the `error` is populated and handle it accordingly:

- https://github.com/WordPress/wordpress-develop/blob/17415a747166e0c90bcf8417d689ad9c556fb90b/src/wp-includes/fonts.php#L178
- https://github.com/WordPress/wordpress-develop/blob/17415a747166e0c90bcf8417d689ad9c556fb90b/src/wp-includes/rest-api/endpoints/class-wp-rest-font-faces-controller.php#L869
- https://github.com/WordPress/wordpress-develop/blob/17415a747166e0c90bcf8417d689ad9c556fb90b/src/wp-includes/rest-api/endpoints/class-wp-rest-font-faces-controller.php#L869

matiasbenedetto committed:
Merge branch 'trunk' into add/font-dir-default-non-persisted

azaozz commented:
Trying an alt solution in https://github.com/matiasbenedetto/wordpress-develop/pull/1. It simplifies the code a bit and only depends on `wp_is_writable( WP_CONTENT_DIR )` to determine whether to move the /fonts dir or not.

Also has an optional param to tell `wp_get_font_dir()` whether to actually make the directory or just get the path/data for it (works similarly to wp_upload_dir() `$create_dir` param).

Also see https://github.com/WordPress/wordpress-develop/pull/6211#issuecomment-2005459537 re: an idea how to "fix" the lambda filter/possible loop for wp_upload_dir() in _wp_handle_upload() by adding an "override".

peterwilsoncc commented:
@azaozz The alternative PR in to this branch doesn't work as it can place font files in `uploads/fonts` on the first install and `wp-content/fonts` on the second install. If the content directory doesn't exist initially for an upload it's created. 

See the gif attached (in my local setup `wp-content` is simply `content`)

![different-dirs](https://github.com/WordPress/wordpress-develop/assets/519727/cfa6cdaf-26c0-40e5-910d-9e6529e34501)

It also causes further lock in to having to work around the infinite loop bug rather than fix it.

azaozz commented:
> it can place font files in uploads/fonts on the first install and wp-content/fonts on the second install

You mean `WP_CONTENT_DIR` was non-writable on the first try to install a font and suddenly became writable when trying to install a font for the second time? Otherwise how can WP ever make directories or add files to a non-writable directory? 

Would it be possible to try couple of `var_dump( is_writable( WP_CONTENT_DIR ) );`?
(Btw, having quite a bit of problems with PHP's `chmod()` when trying to test this. Doesn't seem to work at all. Could that be related?)

> It also causes further lock in to having to work around the infinite loop bug rather than fix it.

But the fix for that infinite loop is to not use a filter there at all, just do a simple "override" like the rest of `_wp_upload_dir()`. Imho there is no good reason for that lambda filter anyway. Discussion: https://github.com/WordPress/wordpress-develop/pull/6211#issuecomment-2005459537.

azaozz commented:
Made the alt patch into a stand-alone PR (as requested): https://github.com/WordPress/wordpress-develop/pull/6292

peterwilsoncc commented:
Closing this PR off per https://make.wordpress.org/core/2024/03/21/font-library-update-storage-of-font-files/. Thanks everyone for your work on this, I appreciate it.
